anywidget/_descriptor.py

```diff
@@ -553,11 +560,11 @@ def _connect_psygnal(obj: object, send_state: Callable) -> Callable | None:
     if events is not None:
 
         @events.connect
-        def _on_psygnal_event(event: psygnal.EmissionInfo):
+        def _on_psygnal_event(event: psygnal.EmissionInfo) -> None:
             send_state({event.signal.name})
 
-        def _disconnect():
-            events.disconnect(_on_psygnal_event)
+        def _disconnect() -> None:
+            cast("psygnal.SignalGroup", events).disconnect(_on_psygnal_event)
```

Comment:
yeah, the function `cast` just returns the second value (`cast = lambda x: x[1]`) ... but it will attempt to evaluate the first object, even if `from __future__ import annotations` is used (in that way, I guess `cast` is a little different than all of the other uses of types in annotations... this isn't really an "annotation" per se)

as an example, this file would import just fine, but fail to call the function:
```py
from __future__ import annotations
from typing import cast

def f():
    cast(notimported, 1)

if __name__ == '__main__':
    print("about to run")
    f()
```

an equally valid approach here is to simply add a `# type: ignore` to that line  (the problem is that it still thinks that `events` _might_ be `None` ... and don't quite know why (it appears the type narrowing in our conditional above isn't making it into the scope of the `_disconnect` function)

Reply:
Also totally fine with this. Thanks for the explanation!

> and don't quite know why (it appears the type narrowing in our conditional above isn't making it into the scope of the _disconnect function)

I'm guessing it's because we've said that event could be `None`, and because it's within a closure, it _could_ be `None` in the future... (even though we know it won't be). I bet binding `events` as a function parameter would make the type checker happy:

```python

        def _disconnect(events = events) -> None:
            events.disconnect(_on_psygnal_event)
```

I ran into something like this the other day with for-loops I think:

```
funcs = []
for i in range(4):
  def log():
    print(f"count is {i}")
  funcs.append(i)

for fn in funcs:
  fn()
# count is 3
# count is 3
# count is 3
# count is 3
```

Reply:
> I bet binding events as a function parameter would make the type checker happy:

that was the first thing I tried too! surprisingly didn't work, and I gave up :joy:

Reply:
lol, ok beyond my knowledge. let's go for the cast then haha